### ISSUE TITLE
Dynamic Provider can not find authentication scheme?

### ISSUE THREAD

**Issue by kkdeveloper7** [DuendeSoftware/Support#1260]
**Which version of Duende IdentityServer are you using?**
7

**Which version of .NET are you using?**
dotnet8

**Describe the bug**
A little background on an issue. We are planning to use multiple identity providers in our application. Many of them will be using SAML2. After Implementing `IIdentityProviderStore` and registering with
`AddIdentityProviderStore<CustomDynamicProviderStore>()`  I am getting scheme not found exception. 

Do I still need to register Authentication schemes manually or I can use `AddOpenIdConnect` and `AddSaml2`? It looks that I need to still register Handlers manually? 

Source: https://docs.duendesoftware.com/identityserver/v7/ui/login/dynamicproviders/

**Comment by AndersAbel** [comment]
@kkdeveloper7 I think this is already sorted out through support, is it?

**Comment by kkdeveloper7** [comment]
> @kkdeveloper7 I think this is already sorted out through support, is it?

Yes, thank you.

**Comment by attilah** [comment]
@kkdeveloper7 @AndersAbel Would you share what was the solution in your case as we are just at the point where after auth we hit the same error ASP.Net error page that says can't find idp_xxxx scheme.

**Comment by kkdeveloper7** [comment]
> @kkdeveloper7 @AndersAbel Would you share what was the solution in your case as we are just at the point where after auth we hit the same error ASP.Net error page that says can't find idp_xxxx scheme.

Are you having issues registering saml2? Or getting dynamic providers working in general?

**Comment by attilah** [comment]
Dynamic providers, OIDC, no SAML2 for us.

**Comment by kkdeveloper7** [comment]
> Dynamic providers, OIDC, no SAML2 for us.
> 
> My suspicion is that somehow the dynamic registration is not correct as the redirect uri is not pointing to the /federated prefix but to the concrete IDP (static) url pattern, and perhaps that's the root cause, but can only confirm it Tomorrow.

First you need to implement this interface `TImplementation : IIdentityProviderStore` and register with Identity Server `AddIdentityProviderStore<TImplementation>()`

Inside of `public async Task<IdentityProvider?> GetBySchemeAsync(string scheme)` make sure you return implementation of `OidcProvider` and do not override `Type` Property. This property is going to be set by default to `oidc` that is going to get mapped to `NamedOptions` later I believer. 

Next implement this interface `IConfigureNamedOptions<OpenIdConnectOptions>`, I dont believe you need to register this as they taking care of this mapping. What this will do, it will invoke `public void Configure(string? name, OpenIdConnectOptions options)` method every time `OpenIdConnectOptions` object is needed in constructor. The `name` parameter will container you `scehme` name that was configured in previous step when you mapped OidcProvider object. 
 
Once all of this is complete when you go though your flow, Challenge the scheme that you configured prior. Make sure that scheme exists. 

```
return Challenge(new AuthenticationProperties({your_props})
                {
                   ...
                }, {scheme_name});
```

`IConfigureNamedOptions<OpenIdConnectOptions>` this will eb invoked on the way out when you challenge and on the way in where you set your redirect urls are

`{application_host}/federation/{scheme_name}/signin` this should be your default redirect unless overridden.

**Comment by attilah** [comment]
@kkdeveloper7 thanks for the elaborate steps, we did all by the books and we did not need the options at the end. As it turned out the real error was hidden behind the scenes in the form of #2523 the backchannel code exchange for token was failing as the token endpoint was null. I open a separate issue to track that, incompatible MSFT Identity assemblies causing this.

Now we're all good!
